File: 232_queue_using_stack.py
class MyQueue:
    """
    // Time Complexity :
                O(n) We need to copy all elements from stack to another
    // Space Complexity :
                O(n) The number of elements in the stack
    // Did this code successfully run on Leetcode :
                Yes
    // Any problem you faced while coding this :
                I initially got confused with the index for the peek operation.
    """
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.in_stack = []
        self.out_stack = []

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.in_stack.append(x)

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        if self.out_stack:
            return self.out_stack.pop()
        else:
            while self.in_stack:
                self.out_stack.append(self.in_stack.pop())
        return self.out_stack.pop()

    def peek(self) -> int:
        """
        Get the front element.
        """
        if self.out_stack:
            return self.out_stack[-1]
        elif self.in_stack:
            return self.in_stack[0]

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return not self.in_stack and not self.out_stack

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: 232-queueUsingStack.java
//Overall Time Complexity: O(1) - Amortized
//Overall Space Complexity: O(N) - Auxillary memory to store elements
class MyQueue {
    Stack<Integer> in;
    Stack<Integer> out;

    public MyQueue() {
        in = new Stack<>();
        out = new Stack<>();
    }

    //Time Complexity: O(1)
    public void push(int x) {
        in.push(x);
    }

    //Time Complexity: O(1) - Amortized
    public int pop() {
        if(out.isEmpty()){
            shift();
        }
        if(out.isEmpty()) return -1;
        else return out.pop();
    }

    //Time Complexity: O(1) - Amortized
    public int peek() {
        if(out.isEmpty()){
            shift();
        }
        return out.peek();
    }

    //Time Complexity: O(1)
    public boolean empty() {
        if(out.isEmpty() && in.isEmpty()) return true;
        else return false;
    }

    //Time Complexity: O(N) - Expensive but happens only some times
    private void shift(){
        while(!in.isEmpty()){
            out.push(in.pop());
        }
        return;
    }
}
